common/models/transducer/rna_pure_tf.py

```diff
@@ -0,0 +1,641 @@
+      # last dimension is (state-idx, label-idx)
+      # sel_blank = np.stack([u_ranged, np.tile(blank_tiled, [1, max_len])], axis=-1)  # (B,)
+      # sel_emit = np.stack([u_ranged_shifted, labels_emit], axis=-1)  # (1,U|n) | (B, U|n)-> (B, n, 2)
+      # sel_same = np.stack([u_ranged, labels_same], axis=-1)
+      # sel = np.where((argmax_idx == 0)[..., np.newaxis],
+      #                sel_blank,  # blank
+      #                np.where((argmax_idx == 1)[..., np.newaxis],
+      #                         sel_emit,  # emit
+      #                         sel_same))  # same
+      blank_tiled = tf.tile([[blank_index]], [n_batch, 1])  # (B, 1)
+      labels_exp = tf.concat([labels, blank_tiled], axis=1)  # (B, U+1)
+      # label_idxs = tf.stack([
+      #   tf.range(n_batch),
+      #   tf.range(max_len)-1,
+      # ], axis=-1)  # (B, 2)
+      b, r = tf.meshgrid(
+        tf.range(n_batch),
+        tf.maximum(0, tf.range(max_len) - 1),
+        indexing='ij'
+      )
+      label_idxs = tf.stack([b, r], axis=-1)
+      labels_emit = tf.gather_nd(labels_exp, label_idxs)  # (B, n)  labels[u-1]
+      labels_same = labels_emit
+      sel_blank = tf.stack([u_ranged, tf.tile(blank_tiled, [1, max_len])], axis=-1)
+      sel_emit = tf.stack([u_ranged_shifted, labels_emit], axis=-1)  # (B, n) | (B,)
+      sel_same = tf.stack([u_ranged, labels_same], axis=-1)
+
+      argmax_idx_tiled = tf.tile(argmax_idx[..., tf.newaxis], [1, 1, 2])  # (B, n, 2)
+      sel = tf.where(tf.equal(argmax_idx_tiled, 0),
+                     sel_blank,  # blank
+                     tf.where(tf.equal(argmax_idx_tiled, 1),
+                              sel_emit,  # emit
+                              sel_same))  # same
+      # we need to pad so we can stack the TA later on (instead of triangular shape)
+      sel_padded = tf.pad(sel, [[0, 0], [0, max_target - max_len], [0, 0]])
+      bt_ta = bt_ta.write(n - 1, sel_padded)
+    else:
+      bt_ta = None
+
+    red_op = tf.stack(reduction_args, axis=0)  # (2, B, N)
+    red_op = py_print_iteration_info("red-op", red_op, n, debug=debug)
+    new_column = tf.math.reduce_logsumexp(red_op, axis=0)  # (B, N)
+
+    new_column = new_column[:, :n]
+    new_column = py_print_iteration_info("new_column", new_column, n, debug=debug)
+    ret_args = [n + 1, alpha_ta.write(n, new_column)]
+    return ret_args + [bt_ta] if with_alignment else ret_args
+
+  n = tf.constant(2)
+  initial_vars = [n, alpha_ta]
+  if with_alignment:
+    initial_vars += [bt_ta]
+  final_loop_vars = tf.while_loop(cond, body_forward,
+                                  initial_vars,
+                                  parallel_iterations=1,  # iterative computation
+                                  name="rna_loss")
+  if with_alignment:
+    final_n, alpha_out_ta, bt_ta = final_loop_vars
+    bt_mat = tf.transpose(bt_ta.stack(), [1, 0, 2, 3])  # (T, B, U, 2) -> (B, T, U, 2)
+    alignments = compute_alignment_tf(bt_mat, input_lengths, label_lengths + 1)
```

Comment:
>No? This pretty much looks like it calculates the alignment. Also the docstring says this.

Yes, but I am not sure if we can totally split them as you see here. It takes as input information from the forward pass.

Reply:
You can always just do `from .rna_alignment_pure_tf import compute_alignment_tf`. There is no problem with that.

But anyway, now that I see this, I wonder if this is actually correct... Why does the alignment needs this `bt_mat`? It should directly operate on the posteriors. There might be some bug here. You should keep that in mind, and check that at some later point (so far we would just overtakes Andres logic/code, and just rearange/reorganize).

The alignment code and the forward code could actually share common code. You go from left to right (over axis t) or diagonal (over axis u), and in one case you sum the probabilities, in the other case, you take the max. So the common code could switch between sum or max, and is mostly the same. (You know how this dynamic programming over such a lattice over possible alignments works, right?)

Ok, but for the alignment, you need to remember the argmax you use in each frame as well, and need to backtrack that by hand, so this is more complicated.

Ok, maybe this is actually what `compute_alignment_tf` does here? So then it might make sense to keep that in the same file.

Maybe just leave the file as-is for now, but still, call it `rna_forward_align_pure_tf` or maybe `rna_align_sum_max_pure_tf` or sth like that.

Reply:
>But anyway, now that I see this, I wonder if this is actually correct... Why does the alignment needs this bt_mat? It should directly operate on the posteriors. There might be some bug here. You should keep that in mind, and check that at some later point (so far we would just overtakes Andres logic/code, and just rearange/reorganize).

`compute_alignment_tf ` just takes the backtrack matrix and gives out the alignment. Somehow forward loss matrix(sum) and forward alignment matrix(max) are computed in the same function. In some point, we should maybe separate them.

>The alignment code and the forward code could actually share common code. You go from left to right (over axis t) or diagonal (over axis u), and in one case you sum the probabilities, in the other case, you take the max. So the common code could switch between sum or max, and is mostly the same. (You know how this dynamic programming over such a lattice over possible alignments works, right?)

So yes, you go over any of the up-left to down-right diagonal of the **product** of the forward and backward matrix at that diagonal and take the sum for the loss. For the alignment if I understand it right one has to take the max every step while calculating the forward and backward matrix and at the end take the max over the diagonal we decide to unite(multiply) the backward and forward matrixes  once more.
But that is done when you keep both forward and backward matrixes. 

I see here only the forward alpha matrix. At this [part](https://github.com/rwth-i6/returnn-experiments/pull/61/files/e0d720998495bd695ffa20717b37fc524b20462c#r611086134), the `alpha[T,N]` is used for the loss. 


>Ok, but for the alignment, you need to remember the argmax you use in each frame as well, and need to backtrack that by hand, so this is more complicated.
Ok, maybe this is actually what compute_alignment_tf does here? So then it might make sense to keep that in the same file.
Exactly.

>Maybe just leave the file as-is for now, but still, call it rna_forward_align_pure_tf or maybe rna_align_sum_max_pure_tf or sth like that.

`rna_align_sum_max_pure_tf ` this sounds more appropriate.

Reply:
I'd say `rna_forward_sum_max_pure_tf` would be actually a better name for `tf_forward_shifted_rna`.

Reply:
> compute_alignment_tf just takes the backtrack matrix and gives out the alignment. Somehow forward loss matrix(sum) and forward alignment matrix(max) are computed in the same function.

Ok, this is then all good, and exactly as it is optimal already.

> In some point, we should maybe separate them.

No, in this case, it is good to have them together, as it is really the same code (only diff is sum vs max), and alignment need the special backtracking logic.